microsoft/LightGBM
# [ci] fix linux runners running out of disk space (fixed #6635)

fixes #6635

It seems that CI jobs on the self-hosted linux runner pool in Azure DevOps are failing because those runners don't have enough free disk space.

I found that around 55% of the disk space on those runners is occupied by no-longer-user container images (https://github.com/microsoft/LightGBM/issues/6635#issuecomment-2323815123). This proposes introducing one new CI job that deletes images that are more than 30 days old and not currently in usage by any containers.

## Notes for Reviewers

### Why not run this cleanup on every CI job?

It'd add around 30-45 seconds to every run of every Azure DevOps linux CI job.

I don't think that's necessary... running this cleanup once per CI *run* (one one randomly-assigned runner in the self-hosted pool) should hopefully be enough to prevent the disk space from filling up way again.

I'm not sure how many total runners there are in the pool introduced in #6407 ... @shiyu1994 could you tell us? That would help with understanding how many runs might be required to clean up every node at least once.

### How to test this

On any Azure DevOps run, check the `warnings` tab

<img width="1430" alt="Screenshot 2024-09-02 at 12 20 28 AM" src="https://github.com/user-attachments/assets/a84ab8a2-657a-44d3-b74a-845058053601">

([example build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16981&view=results))

Over the next few days, we should see the number of such warnings decrease... and eventually see 0 warnings related to disk usage on the linux runners.

Also check the output of the `Maintenance` job.

<img width="1032" alt="Screenshot 2024-09-02 at 12 22 27 AM" src="https://github.com/user-attachments/assets/c038fa2e-074a-43b3-9f97-d2f571eec674">

([example build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16981&view=logs&j=109b04a8-e507-5fc5-3d1a-042c6dd15986&t=6221f19d-281f-5be7-2f83-a42abd4fa0df))

The end of those runs should show only 2 container images on the the host... the one used on the `Linux` jobs and the ones used on the `Linux_latest` jobs.

Over the next few days, we should see log messages like "Total reclaimed space: 14.12GB". Those should eventually stop showing up, as all the runners are cleaned up.

### So is this job temporary?

No, I'm proposing this as a permanent addition to LightGBM's CI.

That way, we'll automatically be protected against disk-space issues in situations like the following:

* switching to a new image for `ubuntu-latest` (via new Ubuntu version or just security patches pushed to the official Ubuntu images)
* changes to the manylinux image (e.g. via new commits to https://github.com/guolinke/lightgbm-ci-docker)
